pandas/core/interval.py

```diff
@@ -0,0 +1,484 @@
+import operator
+
+import numpy as np
+
+from pandas.core.base import PandasObject, IndexOpsMixin
+from pandas.core.common import (_values_from_object, _ensure_platform_int,
+                                _get_next_label, _get_prev_label)
+from pandas.core.index import (Index, _ensure_index, default_pprint,
+                               InvalidIndexError, MultiIndex)
+from pandas.lib import (Interval, IntervalMixin, IntervalTree,
+                        interval_bounds_to_intervals,
+                        intervals_to_interval_bounds)
+from pandas.util.decorators import cache_readonly
+import pandas.core.common as com
+
+
+_VALID_CLOSED = set(['left', 'right', 'both', 'neither'])
+
+
+def _get_interval_closed_bounds(interval):
+    """
+    Given an Interval or IntervalIndex, return the corresponding interval with
+    closed bounds.
+    """
+    left, right = interval.left, interval.right
+    if interval.open_left:
+        left = _get_next_label(left)
+    if interval.open_right:
+        right = _get_prev_label(right)
+    return left, right
+
+
+class IntervalIndex(IntervalMixin, Index):
+    """
+    Immutable Index implementing an ordered, sliceable set. IntervalIndex
+    represents an Index of intervals that are all closed on the same side.
+
+    .. versionadded:: 0.18
+
+    Properties
+    ----------
+    left, right : array-like (1-dimensional)
+        Left and right bounds for each interval.
+    closed : {'left', 'right', 'both', 'neither'}, optional
+        Whether the intervals are closed on the left-side, right-side, both or
+        neither. Defaults to 'right'.
```

Comment:
Why did you default to intervals that are closed on the right?

[numpy.histogram](http://docs.scipy.org/doc/numpy-dev/reference/generated/numpy.histogram.html) and other histogram software I know has intervals that are closed on the left, so that seems like the better default to me for use cases where this is used to represent [histograms](https://en.wikipedia.org/wiki/Histogram).

Feel free to ignore this comment if histograms are not the main or a major use case for this and right-closed intervals are more natural for your applications. I'm looking forward to the docs or example notebook on how to use this ... thanks for implementing this!